File: hw/chisel/src/main/scala/Frontend/BP/BP.scala
/* ------------------------------------------------------------------------------------
* Filename: BP.scala
* Author: Hakam Atassi
* Date: Apr 26 2024
* Description: A wrapper for the primary branch prediction structures (ie, RAS, BTB, GSHARE, GHR)
* License: MIT
*
* Copyright (c) 2024 by Hakam Atassi
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
* ------------------------------------------------------------------------------------ 
*/

package ChaosCore

import chisel3._
import circt.stage.ChiselStage 
import chisel3.util._
import java.io.{File, FileWriter}
import java.rmi.server.UID

class BP(coreParameters:CoreParameters) extends Module{
    import coreParameters._
    val io = IO(new Bundle{
        // Flush
        val flush       = Input(Bool())

        // Predict Channel
        val predict     = Flipped(Decoupled(new frontend_memory_request(coreParameters)))    // PC input

        // Commit Channel 
        val commit      = Flipped(ValidIO(new commit(coreParameters)))

        // Revert Channel
        val RAS_update  = Input(new RAS_update)

        // RAS Channel
        val RAS_read    = Output(new RAS_read(coreParameters))

        // GHR Channel
        val GHR         = Input(UInt(GHRWidth.W))

        // Prediction Channel (output)
        val prediction  = Decoupled(new prediction(coreParameters))    // Output of predictions
    })

    ////////////////////
    // OUTPUT BUNDLES //
    ////////////////////
    val prediction = Wire(Decoupled(new prediction(coreParameters)))

    /////////////
    // MODULES //
    /////////////
    val gshare      = Module(new gshare(coreParameters)) // FIXME: should this be addressed by the PC or by the fetch packet aligned PC?
    val BTB         = Module(new hash_BTB(coreParameters))
    val RAS         = Module(new RAS(coreParameters))

    ///////////////
    // GHR LOGIC //
    ///////////////
    val misprediction       = Wire(Bool());                         misprediction := 0.U

    val misprediction_TOS   = Wire(UInt(log2Ceil(RASEntries).W));   misprediction_TOS := 0.U
    val misprediction_NEXT  = Wire(UInt(log2Ceil(RASEntries).W));   misprediction_NEXT := 0.U

    misprediction       :=  io.commit.bits.is_misprediction

    misprediction_TOS   :=  io.commit.bits.TOS
    misprediction_NEXT  :=  io.commit.bits.NEXT

    //////////////////
    // Commit logic //
    //////////////////
    // PHT => updates on cond branches, correct or incorrect
    // BTB => updates on truely taken branches
    // GHR => updates speculatively. Reverts on reverts from predecoder or mispredictions
    // RAS => updates speculatively. Reverts on mispredictions

    val update_BTB = Wire(Bool())
    val update_PHT = Wire(Bool())
    val update_RAS = Wire(Bool())

    // BTB and PHT are updated for ALL taken branches
    // Regardless of type (BT, JAL, etc...)
    update_BTB := io.commit.bits.T_NT && io.commit.valid
    update_PHT := (io.commit.bits.br_type =/= br_type_t.NONE) && io.commit.valid
    update_RAS := !misprediction


    /////////////////
    // Init gshare //
    /////////////////
    // predict port
    gshare.io.predict_GHR               := io.GHR
    gshare.io.predict_PC                := io.predict.bits.addr
    gshare.io.predict_valid             := io.predict.valid

    // commit port
    gshare.io.commit                <> io.commit
    gshare.io.commit.valid          := update_PHT

    //////////////
    // Init BTB //
    //////////////
    // Reminder: BTB only updates on taken branches...
    // predict port
    BTB.io.predict_PC                       := io.predict.bits.addr
    BTB.io.predict_valid                    := io.predict.valid

    // commit port
    BTB.io.commit           <>  io.commit
    BTB.io.commit.valid     :=  update_BTB

    ///////////////////////////////
    // Init Return-Address-Stack //
    ///////////////////////////////
    // handle misprediction
    RAS.io.revert_NEXT  :=   misprediction_NEXT
    RAS.io.revert_TOS   :=   misprediction_TOS
    RAS.io.revert_valid :=   misprediction

    // update port
    RAS.io.wr_addr      := io.RAS_update.call_addr
    RAS.io.wr_valid     := io.RAS_update.call && update_RAS
    RAS.io.rd_valid     := io.RAS_update.ret  && update_RAS

    // Assign outputs

    // RAS
    io.RAS_read.ret_addr := RAS.io.ret_addr
    io.RAS_read.NEXT     := RAS.io.NEXT
    io.RAS_read.TOS      := RAS.io.TOS

    // BTB
    // FIXME: something needs to be added here to ensure that predictions after 
    // A replated mispredicted packet are correct...

    prediction.bits.target    := BTB.io.BTB_output.BTB_target
    prediction.bits.br_type   := BTB.io.BTB_output.BTB_br_type
    prediction.bits.hit       := BTB.io.BTB_hit
    prediction.bits.T_NT      := gshare.io.T_NT

    prediction.valid        := RegNext(io.predict.valid && !io.flush) //BTB.io.BTB_output.valid && gshare.io.valid
    prediction.ready        := io.prediction.ready

    io.prediction <> prediction


    /////////////////
    // SKID BUFFER //
    /////////////////
    val prediction_skid_buffer      = Module(new Queue(new prediction(coreParameters), 1, flow=true, hasFlush=true, useSyncReadMem=false))

    prediction_skid_buffer.io.enq                  <> prediction
    prediction_skid_buffer.io.deq                  <> io.prediction
    prediction_skid_buffer.io.flush.get            := io.flush

    io.predict.ready        := io.prediction.ready

}

File: py/models/BP/BP.py
from Gshare import *
from HashBTB import *
from RAS import *


class commit:
    def __init__(self, commit_data):
        self.address, self.commit_GHR, self.T_NT, self.target, self.br_type, self.br_mask, self.misprediction, self.TOS, self.NEXT, self.valid = commit_data
    def get(self):
        return (self.address, self.commit_GHR, self.T_NT, self.target, self.br_type, self.br_mask, self.misprediction, self.TOS, self.NEXT, self.valid)
    @staticmethod
    def null():
        return commit((0, 0, 0, 0, 0, 0, 0, 0, 0, 0))
    def __eq__(self, other):
        if isinstance(other, commit):
            return self.get() == other.get()
        return False

class predict:
    def __init__(self, predict_data):
        self.address, self.valid = predict_data
    def get(self):
        return (self.address, self.valid)
    @staticmethod
    def null():
        return predict((0, 0))
    def __eq__(self, other):
        if isinstance(other, prediction):
            return self.get() == other.get()
    def print(self):
        print(self.get(), end=" ")

class revert:
    def __init__(self, revert_data):
        self.address, self.revert_GHR, self.valid = revert_data
    def get(self):
        return (self.address, self.revert_GHR, self.valid)
    @staticmethod
    def null():
        return revert((0, 0, 0))
    def __eq__(self, other):
        if isinstance(other, prediction):
            return self.get() == other.get()
    def print(self):
        print(self.get(), end=" ")

class RAS_update:
    def __init__(self, RAS_update_data):
        self.call, self.ret, self.ret_addr = RAS_update_data
    def get(self):
        return (self.call, self.ret, self.ret_addr)
    @staticmethod
    def null():
        return RAS_update((0, 0, 0))
    def __eq__(self, other):
        if isinstance(other, prediction):
            return self.get() == other.get()
    def print(self):
        print(self.get(), end=" ")

class prediction:
    def __init__(self, prediction_data):
        self.hit, self.target, self.br_type, self.br_mask, self.T_NT, self.GHR, self.valid = prediction_data
    def get(self):
        return (self.hit, self.target, self.br_type, self.br_mask, self.T_NT, self.GHR, self.valid)
    @staticmethod
    def null():
        return prediction((0,0,0,0,0,0,0))
    def __eq__(self, other):
        if isinstance(other, prediction):
            return self.get() == other.get()
    def print(self):
        print(self.get(), end=" ")


class BP:
    def __init__(self, RAS_entries=128, BTB_entries=4096, GHR_width=16, fetch_width=4):
        self.RAS = RAS(entries=RAS_entries)
        self.BTB = HashBTB(entries=BTB_entries, fetch_width=fetch_width)
        self.PHT = PHT(GHR_width=GHR_width)
        self.GHR = GHR(GHR_width=GHR_width)

    #####################
    # BP INPUT CHANNELS #
    #####################

    def commit(self, commit_msg):
        if(commit_msg.valid == 0): return 
        ## commits update the BTB and PHT with branch information
        address, commit_GHR, T_NT, target, br_type, br_mask, misprediction, TOS, NEXT, valid = commit_msg.get()
        if(T_NT):
            self.BTB.commit(address=address, target=target, br_type=br_type, br_mask=br_mask)
        if(br_type == 0):
            self.PHT.commit(PC=address, GHR=commit_GHR, resolved_direction=T_NT)

        if(int(misprediction)==1):
            ## revert RAS and GHR
            self.GHR.write(commit_GHR) # revert GHR
            self.RAS.revert(NEXT, TOS)
    
    def predict(self, predict_msg):
        ## predict requests read the BTB and PHT. RAS output is assumed already available.
        ## They also speculatively update the GHR based on the predicted taken direction (for branches only)
        # Prediction consists of BTB+PHT+Current GHR
        
        if(predict_msg.valid == 0): return
        address, valid = predict_msg.get()

        predict_GHR=self.GHR.read()
        BTB_resp = self.BTB.predict(address=address)
        PHT_resp = self.PHT.predict(PC=address, GHR=predict_GHR)
        if(BTB_resp[0] and BTB_resp[2]==0): # only update GHR when BTB indicates a branch
            self.GHR.update(PHT_resp)
        return prediction((*BTB_resp, PHT_resp, predict_GHR, valid))

    def revert(self, revert_msg):
        if(revert_msg.valid == 0): return
        ## reverts update the GHR only
        # reverts are mini mispredictions. They essentialy redirect the frontend before it has the change to
        # incorrectly update things like the RAS
        address, revert_GHR, valid = revert_msg.get()
        self.GHR.write(revert_GHR)

    def send_all_requests(self, predict, commit, revert, RAS_update):
        is_misprediction   = commit.misprediction
        is_revert          = revert.valid

        #if(commit.valid): print(f"model commit {hex(commit.address)}")

        #print(f"model: {commit.valid} {hex(commit.address)} {hex(commit.misprediction)}")

        if(is_misprediction and commit.valid):
            self.commit(commit)  # on misprediction, handle misprediction first]
            # on mispredictions, ignore reverts even if they are valid
        elif(is_revert):
            # FIXME: is this order correct??
            self.revert(revert)
            self.commit(commit)
        else: # normal operation
            self.commit(commit)
            self.RAS_update(RAS_update)

        prediction = None
        if(predict.valid):
            prediction = self.predict(predict)

        return prediction

    ########################
    # SUBMODULE INTERFACES #
    ########################

    # RAS
    def RAS_read(self):
        # Get access to current RAS values (without updating)
        RAS_output = self.RAS.get_outputs()
        return RAS_output

    def RAS_update(self, RAS_update_msg):
        call, ret, call_addr = RAS_update_msg.get()
        if call:
            self.RAS.push(addr=call_addr)
        elif ret:
            self.RAS.pop()


    def get_PHT_state(self):
        return self.PHT.get_state()


    # GHR
    def get_GHR(self):
        return self.GHR


